modules/ocf/files/ssl/ocf-lets-encrypt

```diff
@@ -0,0 +1,253 @@
+
+    3. If the answer to (1) or (2) is "no", we invoke acme-tiny to
+       acquire/renew a cert for the domain.
+
+Sounds simple enough? Well, there's more to it. acme-tiny proves that we
+actually own the domain by writing a challenge file and expecting it to
+be served under http://<fqdn>/.well-known/acme-challenge/. This presents
+two problems: 1) on hosts that don't run a web server (e.g. mail), we
+need something to serve that file, and 2) on hosts that do run a web
+server, we may need to acquire the cert before the web server can even
+start.
+
+Our solution is to enforce the following assumption: if there is a
+running web server, it MUST serve up the challenge directory at
+http://<fqdn>/.well-known/acme-challenge. This simplifies our logic to
+the following:
+
+    1. If there's something already listening on port 80, just make any
+       needed certificate requests and expect them to succeed; the
+       running web server will handle it.
+
+    2. If not, then we're either bootstrapping or the host doesn't run a
+       web server. In this case, we start a temporary server (Python's
+       http.server) to serve the challenge files while making requests
+       and shut it down when we're finished.
+
+It is also possible for us to hit our rate limit for Let's Encrypt
+requests, but our limit as of writing is 1000 new certs per week with
+unlimited renewals, so this is only possible in case of catastrophic
+failure (i.e. we lose all of our web vhost certs).
+
+Once the needed certs are acquired, the corresponding .pem and .bundle
+files still have to be generated. This is left to the caller, i.e.
+puppet or a host-specific script.
+
+The return code indicates whether the script did anything, so the caller
+can reload services as necessary:
+
+    0       nothing changed
+    255     certs added/changed
+    else    error
+"""
+import argparse
+import functools
+import os
+import re
+import shutil
+import socket
+import subprocess
+import sys
+import tempfile
+from collections import namedtuple
+from contextlib import contextmanager
+from datetime import datetime
+from datetime import timedelta
+from datetime import timezone
+from pathlib import Path
+
+import dateutil.parser
+
+
+RENEW_WHEN_TIME_REMAINING = timedelta(days=45)
+
+DEFAULT_CERT_STORAGE_DIR = '/etc/ssl/private/'
+LE_ACCOUNT_KEY = Path('/etc/ssl/lets-encrypt/le-account.key')
+ACME_BASE_DIR = Path('/var/lib/lets-encrypt/')
+
+
+Request = namedtuple('Request', ('domain', 'cert_path', 'has_cert'))
+
+
+def debug(*args, **kwargs):
+    pass
+
+
+def expiration_date(cert_path):
+    output = subprocess.check_output((
+        'openssl', 'x509', '-enddate', '-noout',
+        '-in', str(cert_path),
+    ))
+    m = re.match(b'notAfter=(.+)$', output)
+    assert m, output
+    return dateutil.parser.parse(m.group(1).decode('ascii'))
+
+
+@contextmanager
+def make_csr(domain, key):
+    _, csr_path = tempfile.mkstemp()
```

Comment:
Defensive programming practices suggest you should close the fd that `tempfile.mkstemp()` returns, or use `NamedTemporaryFile` instead. We never know when someone will copy this function's code into the code of some long-running process, and all of a sudden that long-running process is leaking fd's.